File: sources/academy/webscraping/scraping_basics_python/exercises/guardian_publish_dates.py
from datetime import datetime

import httpx
from bs4 import BeautifulSoup

url = "https://www.theguardian.com/sport/formulaone"
response = httpx.get(url)
response.raise_for_status()

soup = BeautifulSoup(response.text, "html.parser")

for article in soup.select('#maincontent ul li'):
    title_tag = article.select_one('h3')
    time_tag = article.select_one('time')

    if not title_tag or not time_tag or 'datetime' not in time_tag.attrs:
        continue

    title = title_tag.text.strip()
    date_iso = time_tag['datetime'].strip()
    date = datetime.fromisoformat(date_iso)

    print(f"{title} | {date.strftime('%a %b %d %Y')}")

File: sources/academy/webscraping/scraping_basics_javascript/exercises/guardian_publish_dates.mjs
import * as cheerio from 'cheerio';

const url = 'https://www.theguardian.com/sport/formulaone';
const response = await fetch(url);

if (!response.ok) {
  throw new Error(`HTTP ${response.status}`);
}

const html = await response.text();
const $ = cheerio.load(html);

for (const element of $('#maincontent ul li').toArray()) {
  const $article = $(element);
  const title = $article.find('h3').text().trim();
  const dateAttr = $article.find('time').attr('datetime');

  if (!title || !dateAttr) {
    continue;
  }

  const date = new Date(dateAttr.trim());
  if (Number.isNaN(date.getTime())) {
    continue;
  }

  console.log(`${title} | ${date.toDateString()}`);
}
